### PR TITLE
Bundle

### PR DESCRIPTION
**Pull request by danburkert**
This is a work-in-progress to expand the example application to be a replicated
mutable register. The goal is to have as simple an example as possible while
still demonstrating the power of distributed consensus. When done, the example
will also be useful for integration tests.

### CODE REVIEW

**Comment by Hoverbear** [comment]
Looks great so far!

**Comment by danburkert** [comment]
I've expanded this with a bunch of recent work.

**Comment by Hoverbear** [comment]
@danburkert Looks good to me. Let me know when it's ready for an r+?

**Comment by danburkert** [comment]
It's pretty much ready.  I will have more commits coming eventually, but they aren't necessarily related to these.

**Comment by Hoverbear** [comment]
Ok, I'll review and merge.

File: src/connection.rs [hunk 1]
```diff
@@ -91,7 +91,7 @@ impl Connection {
             kind: ConnectionKind::Unknown,
             addr: addr,
             stream: socket,
-            backoff: Backoff::with_duration_range(8, 10000),
+            backoff: Backoff::with_duration_range(50, 10000),
```
Comment (Hoverbear) [inline src/connection.rs]: It might be good to make these `const` instead of magic numbers.

File: src/messages.rs [hunk 2]
```diff
@@ -224,12 +225,12 @@ pub fn proposal_response_unknown_leader() -> Rc<MallocMessageBuilder> {
     Rc::new(message)
 }
 
-pub fn proposal_response_not_leader(leader_hint: &str) -> Rc<MallocMessageBuilder> {
+pub fn proposal_response_not_leader(leader_hint: &SocketAddr) -> Rc<MallocMessageBuilder> {
```
Comment (Hoverbear) [inline src/messages.rs]: +1 I prefer this.

File: src/replica.rs [hunk 3]
```diff
@@ -760,90 +791,28 @@ mod test {
...
-        elect_leader(leader.clone(), &mut replicas);
-
-        assert!(replicas[leader].is_leader());
-        assert!(replicas[&replica_ids[1]].is_follower());
-        assert!(replicas[&replica_ids[2]].is_follower());
-        assert!(replicas[&replica_ids[3]].is_follower());
-        assert!(replicas[&replica_ids[4]].is_follower());
-        assert!(replicas[&replica_ids[5]].is_follower());
-        assert!(replicas[&replica_ids[6]].is_follower());
+    fn test_election() {
```
Comment (Hoverbear) [inline src/replica.rs]: Much better!

File: src/state.rs [hunk 4]
```diff
@@ -23,6 +24,8 @@ pub enum ReplicaState {
 pub struct LeaderState {
     next_index: HashMap<ServerId, LogIndex>,
     match_index: HashMap<ServerId, LogIndex>,
+    /// Stores in-flight client proposals.
+    pub proposals: VecDeque<(ClientId, LogIndex)>,
```
Comment (Hoverbear) [inline src/state.rs]: Awesome, this is what I was thinking too. A `VecDeque` is a good structure for this.

**Comment by Hoverbear** [comment]
@homu r+

**Pull request merged**